Reworked change_dir to support symlinks

supagu opened the pull request:
Same as #7935
But I put it in a branch at my end so I can work on other stuff (accidentally had it in my master branch):
Original note for reference:

I have rewritten change_dir to support symlinks.
This works by allowing current_dir to diverge from what getcwd deems is the current directory as getcwd resolves symlinks to the actual real directory which we do not want to do.

This is a potential fix for #6539, #6047 and #5700

supagu committed:
Reworked change_dir to support symlinks

akien-mga commented on drivers/unix/dir_access_unix.cpp:
```diff
@@ -222,36 +222,37 @@ Error DirAccessUnix::make_dir(String p_dir) {
 
-	chdir(current_dir.utf8().get_data()); //ascii since this may be unicode or wathever the host os wants
-	bool worked = (chdir(p_dir.utf8().get_data()) == 0); // we can only give this utf8
+	if (p_dir == ".") {
```
I guess that could go to the top of the function?

akien-mga commented:
The new version looks more readable than the previous one, so if it does not introduce regressions, LGTM. It needs some extensive testing though to ensure that everything works as it should.

supagu commented:
I've been running with it live... and depending on it for months in godot 3. But of course, the more testing the better!

akien-mga commented:
Yeah it's just that it's a quite touchy core API, so I prefer to be 100% sure it's working as it should ;)
In particular, it needs to be tested on all UNIX platforms, so everything !Windows.

akien-mga commented:
Since noone is testing, let's use a different approach.

I'll merge it, and if issues are found, we can either fix them or revert that commit.

#TestInProduction

reduz commented:
Sorry, this had to be reverted. It's an easy way out for a problem that creates more problems. In this case, you can't have two DirAccess instances working together (ie, copy files from one place to another, we use this in a few places in filesystem dock, exporter, etc) because each would change the current path.

If you want to find a real solution, try to find one where two DirAccess instances can work together.
I know this is not an easy problem to solve, and I did not want to spend much time on it because I honestly don't think it's relevant enough.

akien-mga commented:
Also see these [IRC logs](http://godot.eska.me/irc-logs/devel/2017-08-21.log) from 22:55 to see when we discussed this PR.

supagu commented:
Thanks, I'll take a look when i get a chance

supagu commented:
For later reference:

`
[22:55:28] <Akien> !next [22:55:29] <IssueBot> #10469: Switch back to project folder after making data folder | https://git.io/v5vDh [22:55:47] <Akien> Alright, that one might need some discussion [22:56:59] <Akien> Context: I recently merged #8144 which had been sitting around with no testing for months, and since the new code looks IMO clearer I wanted to give it a try [22:57:00] <IssueBot> #8144: Reworked change_dir to support symlinks | https://git.io/vSkM4 [22:57:14] <Akien> (we all know that to get testing in this project you need to merge into master :p) [22:57:27] <Akien> #10469 aims at fixing one apparently regression from this previous PR [22:57:28] <IssueBot> #10469: Switch back to project folder after making data folder | https://git.io/v5vDh [22:58:12] <Akien> So I guess we should re-review #8144 while reviewing #10469 too, since it changes a quite important method [22:58:14] <IssueBot> #8144: Reworked change_dir to support symlinks | https://git.io/vSkM4 [22:58:14] <IssueBot> #10469: Switch back to project folder after making data folder | https://git.io/v5vDh [22:58:32] <Akien> ping supagu ;) [22:58:59] <reduz> DirAccess is meant to never change dir internally [22:59:23] <reduz> you might want to have many of them at the same time [23:00:46] <reduz> Akien: I don't think the original PR you merged is desired either [23:01:29] <reduz> Akien: it should actually be reverted [23:01:37] <reduz> Akien: and closed [23:01:44] <Akien> Yeah, since I couldn't get anyone to test it nor review it, I gave it a chance, but aware that if it introduced issues it could be reverted [23:01:57] <Akien> But it aims at fixing our broken symlink support, so it addresses several bugs [23:02:01] <reduz> it's not that it introduced issues, it's broken from the start [23:02:11] <reduz> it does not allow many DirAccess to work together [23:02:34] <reduz> if you want to for example, copy files from a dir and write in another you can't use it [23:02:41] <Akien> And what if it just readded `chdir(prev_dir.utf8().get_data());` at the end? [23:02:42] <reduz> so the PR is not desired from the begining [23:02:52] <reduz> Akien: then we are back at the begining and the bug is not fixed [23:03:02] <reduz> well, the symlink bug [23:03:20] <reduz> it needs to be fixed in a different way, this is the easy way out and it's broken [23:03:39] <reduz> so #8144 needs to be reverted to begin with [23:03:39] <IssueBot> #8144: Reworked change_dir to support symlinks | https://git.io/vSkM4 [23:04:05] <reduz> Akien: let me know if you want to revert it yourself, or i should do it [23:04:40] <Akien> reduz: Go ahead, but please also give some pointers on the reverted PR about how to properly fix symlink support, if possible at all [23:04:45] <reduz> Akien: also please don't merge PRs related to core without my approval [23:05:00] <reduz> we must have very special care with core because bugs there propagate in strange ways [23:05:05] <Akien> reduz: I've been trying to have your approval for 6 months on this PR.. [23:05:26] <Akien> Nobody ever cared to review it, so I went the only way that I know works: merge, and revert if not good :) [23:05:28] <reduz> Akien: I just didn't have the time to think how to fix this problem properly, it's not easy t all [23:06:07] <AlienCat> strange, after adding a dock, the resource dock dissapear [23:09:15] <reduz> Akien: yeah but in many cases i know the PR is broken but can't think of how to be constructive enought o fix it.. it's not that only that it might not be tested [23:09:39] <TMM> Akien, is there still a bug left for this symlink stuff? I could have a look [23:09:39] <reduz> Akien: so merging it only makes us waste time. I generally try to write in the PR what I think the problem is, but in this case it seems i forgot [23:10:03] <Akien> TMM: Yeah, check the issue linked in the OP [23:10:06] <Akien> *issues [23:10:20] <TMM> OK, I'll try to come up with something [23:10:35] <Akien> There were some nice readability changes in that PR btw, would be nice to readd them without changing the logic :) [23:10:47] <Akien> (if possible at all, I haven't checked thoroughly) [23:11:02] <Akien> !next`

supagu commented:
Okay I sat down and had a poke at fixing this bug (which has been reverted by Reduz in #10518).
@NathanWarden's first fix in #10454 really is the correct fix instead of #10469.

So for this to be merged, #10454 should also be merged along with it.
